src/middlewared/middlewared/plugins/audit/utils.py

```diff
@@ -55,4 +66,100 @@ def generate_audit_table(svc, vers):
     )
 
 
+def parse_query_filters(
+    services: list,
+    filters: list,
+    skip_sql_filters: bool
+) -> tuple:
+    """
+    NOTE: this method should only be called by audit.query
+
+    This method tries to optimize audit query based on provided filter.
+    Optimizations are:
+
+    1. limit databases queried
+    2. generate sql filters where appropriate
+
+    returns a tuple of services that should be queried on backend and validated
+    SQL-safe filters.
+
+    We err on side of caution here since we're dealing with audit results.
+    This means that we skip optimized filters if the field is a JSON one, and
+    do not try to pass disjunctions to sqlalchemy. In future if needed we
+    can loosen these restrictions with appropriate levels of testing and
+    validation in auditbackend plugin.
+    """
+    services_to_check = set(services)
+    filters_out = []
+
+    for f in filters:
+        if len(f) != 3:
+            continue
+
+        if f[0] == 'service':
+            # we are potentially limiting which services may be audited
+
+            if isinstance(f[2], str):
+                svcs = set([f[2]])
+            else:
+                svcs = set(f[2])
+
+            match f[1]:
+                case '=' | 'in':
+                    services_to_check = services_to_check & svcs
+                case '!=' | 'nin':
+                    services_to_check = services_to_check - svcs
+                case _:
+                    # Other filters quite unlikely to be used
+                    # by end-users so we'll just skip optimization
+                    # and rely on filter_list later on
+                    pass
+
+            if not services_to_check:
+                # These filters are guaranteed to have no results. Bail
+                # early and let caller handle it.
+                break;
+
+        if skip_sql_filters:
+            # User has manually specified to pass all these filters to datastore
+            continue
+
+        if f[0] not in SQL_SAFE_FIELDS:
+            # Keys that contain JSON data are not currently supported
+            continue
+
+        filters_out.append(f)
+
+    return (services_to_check, filters_out)
+
+
+def may_use_sql_filters(
+    services: list,
+    filters_in: list,
+    filters_for_sql: list,
+    options: dict
+) -> bool:
+    if filters_in != filters_for_sql:
+        # We will need to do additional filtering after retrieval
+        return False
```

Comment:
In this (and all other cases) we still can use SQL filters, can't we? We'll only pass some of the filters (that are SQL-compatible) to SQLite, it will use indexes as much as possible, then we'll run additional filtering/sorting with Python.

Reply:
Right. I probably should name this better. Filtering gets complicated when dealing with queries such as:
```
[["event_data.credentials.credentials_data.token", "=", "xltISV7M7eBKEw0kOody5nqlC6QZqniGhugX_ZePOOhT42tErGVHyZoRZz-FkgvI"]]
```
Where field in SQL is JSON. SQlite3 actually has JSON support, but conversion of this sort of query is non-trivial and probably not worth our time. This method basically determining whether or not we need to do additional python filtering.